lib/reek/configuration/app_configuration.rb

```diff
@@ -10,135 +14,95 @@ module Configuration
     #
     # @api private
     class AppConfiguration
+      include ConfigurationValidator
       EXCLUDE_PATHS_KEY = 'exclude_paths'
-      attr_reader :exclude_paths, :default_directive, :directory_directives
+      private_attr_writer :directory_directives, :default_directive, :excluded_paths
 
-      # Given this configuration file:
+      # Instantiate a configuration via given path.
       #
-      # ---
-      # IrresponsibleModule:
-      #   enabled: false
-      # "app/helpers":
-      #   UtilityFunction:
-      #     enabled: false
-      # exclude_paths:
-      #   - "app/controllers"
+      # @param path [Pathname] the path to the config file
       #
-      # this would result in the following configuration:
-      #
-      # exclude_paths = [ Pathname('spec/samples/two_smelly_files') ]
-      # default_directive = { Reek::Smells::IrresponsibleModule => { "enabled" => false } }
-      # directory_directives = { Pathname("spec/samples/three_clean_files/") =>
-      #                          { Reek::Smells::UtilityFunction => { "enabled" => false } } }
+      # @return [AppConfiguration]
+      def self.from_path(path: nil)
+        allocate.tap do |instance|
+          instance.instance_eval { find_and_load(path: path) }
+        end
+      end
+
+      # Instantiate a configuration by passing everything in.
       #
+      # @param directory_directives [Hash] for instance:
+      #   { Pathname("spec/samples/three_clean_files/") =>
+      #     { Reek::Smells::UtilityFunction => { "enabled" => false } } }
+      # @param default_directive [Hash] for instance:
+      #   { Reek::Smells::IrresponsibleModule => { "enabled" => false } }
+      # @param excluded_paths [Array] for instance:
+      #   [ Pathname('spec/samples/two_smelly_files') ]
       #
-      # @param options [OpenStruct]
-      #   e.g.: #<OpenStruct report_format =: text,
-      #                      location_format =: numbers,
-      #                      colored = true,
-      #                      smells_to_detect = [],
-      #                      config_file = #<Pathname:config/defaults.reek>,
-      #                      argv = [ "lib/reek/spec" ]>
-      def initialize(options = nil)
-        @directory_directives = {}
-        @default_directive    = {}
-        @exclude_paths        = []
+      # @return [AppConfiguration]
+      def self.from_map(directory_directives: {},
```

Comment:
>> Why not make .from_map take the full contents of a yaml config file as its single parameter, allowing unification of the handling of both cases?

And then we could move it back into the initializer, so abandon factory methods again. And at this point in time, @chastell would swoop in and suggest to make separate factory methods for this. And then we would have an infinite loop :laughing: 

In all seriousness: I believe the two factory methods are fine since they cater to two completely different use cases.
Not to mention that we had it like that before in this PR and it seems just a taaaaaad like bike shedding.:)